cmd/recovery-tool/volmgr/sfdisk.go

```diff
@@ -0,0 +1,260 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+package volmgr
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"os/exec"
+	"strings"
+
+	"github.com/snapcore/snapd/gadget"
+	"github.com/snapcore/snapd/osutil"
+)
+
+var (
+	sectorSize gadget.Size = 512
+
+	_ partitionTool = (*SFDisk)(nil)
+)
+
+// SFDiskDeviceDump represents the sfdisk --dump JSON output format.
+type SFDiskDeviceDump struct {
+	PartitionTable SFDiskPartitionTable `json:"partitiontable"`
+}
+
+type SFDiskPartitionTable struct {
+	Label      string            `json:"label"`
+	ID         string            `json:"id"`
+	Device     string            `json:"device"`
+	Unit       string            `json:"unit"`
+	FirstLBA   uint64            `json:"firstlba"`
+	LastLBA    uint64            `json:"lastlba"`
+	Partitions []SFDiskPartition `json:"partitions"`
+}
+
+type SFDiskPartition struct {
+	Node  string `json:"node"`
+	Start uint64 `json:"start"`
+	Size  uint64 `json:"size"`
+	Type  string `json:"type"`
+	UUID  string `json:"uuid"`
+	Name  string `json:"name"`
+}
+
+type SFDisk struct {
+	device         string
+	partitionTable *SFDiskPartitionTable
+}
+
+func newSFDisk(device string) *SFDisk {
+	return &SFDisk{
+		device: device,
+	}
+}
+
+// DeviceInfo obtains the partitioning and filesystem information from the block
+// device and expresses it as a laid out volume.
+func (sf *SFDisk) DeviceInfo() (*gadget.LaidOutVolume, error) {
+	output, err := exec.Command("sfdisk", "--json", "-d", sf.device).CombinedOutput()
+	if err != nil {
+		return nil, osutil.OutputErr(output, err)
+	}
+
+	var dump SFDiskDeviceDump
+	if err := json.Unmarshal(output, &dump); err != nil {
+		return nil, fmt.Errorf("cannot parse sfdisk output: %v", err)
+	}
+
+	pv, err := positionedVolumeFromDump(&dump)
+	if err != nil {
+		return nil, err
+	}
+
+	// Hold the partition table for later use when creating partitions
+	sf.partitionTable = &dump.PartitionTable
+
+	return pv, nil
+}
+
+// CreatePartitions creates the partitions listed in positionedVolume not listed
+// in usedPartitions. Return a role to device node map.
+func (sf *SFDisk) CreatePartitions(pv *gadget.LaidOutVolume, usedPartitions []bool) (map[string]string, error) {
```

Comment:
We must return device node information here because caller must do further work on the block devices such as filesystem or encrypted device creation. I'm returning a raw map with just the essential information but perhaps something more structured could be nicer.

Reply:
Another idea: keep the node map as an internal state and access it from a new call such as `NodeForRole(role string)`. This looks better than exposing the map itself.